### DISCUSSION TITLE
[Question][Hw5] Some questions regarding specification

### DISCUSSION THREAD

**Discussion by kohs100** [prosyslab-classroom/cs424-program-reasoning#347] [Q&A]
I have few questions about HW5's test cases.

1. Does the function return type always String?
2. Are Start, ntInt, ntConstInt, ntConstStr, ntX nonterminals always exists, and their types are identical to the given example files?
3. Does operator's specification change? (e.g. str.len and str.substr take Start instead of ntX nonterminals)
4. Are there the cases which the function return value is empty string?
5. Are ntConstStr values always singleton string(" ", "A", "B", ...) or empty string("")?
6. For str.indexof (_s_, _t_, _i_), Under my understanding, if _i_ is valid index of _s_ and _t_="", it will just return _i_. However, what value should it return if _i_ is invalid index of _s_ and _t_="" ?

**Comment by sujin0529** [comment]
Hi,

1. Yes, the function return always String type in all test cases.
2. Yes, 5 nonterminals always exist in all example files and their types are identical.
3. No, the specification of the operator does not change.
4. Is it correct to ask if there is a test case with a constraint that the function return value is an empty string? Then there is no such case.
5. If the meaning of "singleton string" means a string of length 1, the ntConstStr value might not be a singleton string. That is, the ntConstStr value can be longer than 1.
6. Since i is an invalid index, the return value is -1.

Please ask me again if I misunderstood the question or if you have any more questions.

Thanks.

**Comment by kohs100** [comment]
Thank you!
